src/dag.rs

```diff
@@ -0,0 +1,483 @@
+    // If we did this, could we typecheck these inputs with the actions? Maybe via
+    // some macro that generates a checker function.
+}
+
+/// A specification for creating a node. This allows dynamic creation of nodes
+/// for subsagas with different instance_ids and creation parameters.
+pub struct NodeSpec<'a> {
+    pub name: &'a str,
+    pub label: &'a str,
+    pub action: &'a str,
+}
+
+/// An abstract description of the shape of node execution
+///
+/// TODO: Do we want to allow arbitrary recursion here?
+pub enum NodeConcurrency<'a> {
+    Linear(Vec<NodeSpec<'a>>),
+    Parallel(Vec<NodeSpec<'a>>),
+}
+
+/// An abstract specification for a subsaga that can be used by a Dag
+/// to generate and append saga nodes.
+///
+/// An example implementation is shown below:
+///
+/// ```ignore
+/// fn create_disk_subsaga<'a>() -> SagaSpec<'a> {
+///     SagaSpec(
+///         // The root node is always the first one. It contains the saga
+///         // parameters.
+///         NodeSpec {
+///             name: "disk_create_params",
+///             label: "ReturnDiskCreateParams",
+///             action: "disk_create_params",
+///         },
+///         vec![
+///             NodeConcurrency::Linear(vec![NodeSpec {
+///                 name: "disk_id",
+///                 label: "GenerateDiskId",
+///                 action: "saga_generate_uuid",
+///             }]),
+///             NodeConcurrency::Parallel(vec![NodeSpec {...}, NodeSpec {...}]),
+///         ],
+///     )
+/// }
+/// ```
+///
+///
+/// TODO: Semantic validation (actions exist in registry, etc...)
+pub struct SagaSpec<'a>(pub NodeSpec<'a>, pub Vec<NodeConcurrency<'a>>);
+
+impl Node {
+    /// Create a new child node for a saga or subsaga
+    pub fn new_child(
+        name: &str,
+        instance_id: u16,
+        label: &str,
+        action: ActionName,
+    ) -> Node {
+        Node {
+            name: name.to_string(),
+            instance_id,
+            label: label.to_string(),
+            action,
+            create_params: None,
+        }
+    }
+
+    /// Create a new root node for a saga or subsaga
+    ///
+    /// The first node in a subsaga is always a root node even though it is
+    /// technically a child node of the outer saga.
+    pub fn new_root<T: ActionData>(
+        name: &str,
+        instance_id: u16,
+        label: &str,
+        action: ActionName,
+        create_params: &T,
+    ) -> Node {
+        let create_params = serde_json::to_value(create_params).unwrap();
+        Node {
+            name: name.to_string(),
+            instance_id,
+            label: label.to_string(),
+            action,
+            create_params: Some(create_params),
+        }
+    }
+}
+
+/// A DAG describing a saga
+//
+// Note: This doesn't implement JSON schema because of Graph and NodeIndex
+// TODO(AJS) - Create a separate metadata type?
+//
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct Dag {
+    pub(crate) name: SagaName,
```

Comment:
Sure. That makes sense.

Reply:
I've changed my mind on this one. I prefer the term "name". I'm not sure that implies load-bearing, but it does imply uniqueness to some degree to me. Label doesn't apply uniqueness and seems more like a "tag". These are just my own preferences, but I'm going to leave it for now if that's ok.

Reply:
I don't believe Steno _does_ assume this value is unique.  The consumer might?  In the case of Omicron, I imagine this will be something like "instance-provision".  It doesn't uniquely identify the execution or the DAG, though it might uniquely identify the purpose or the subsystem that created it?  Maybe some metrics or tooling will assume these values mean something, but I don't think Steno does.

Anyway I don't mind keeping this called "name".  I think we've cleared up the confusion by having newtypes and calling things saga_name vs. node_name vs. action_name.

Reply:
> I don't believe Steno _does_ assume this value is unique. The consumer might? In the case of Omicron, I imagine this will be something like "instance-provision". It doesn't uniquely identify the execution or the DAG, though it might uniquely identify the purpose or the subsystem that created it? Maybe some metrics or tooling will assume these values mean something, but I don't think Steno does.

That's a good point.